server/src/internalClusterTest/java/org/elasticsearch/search/fieldcaps/FieldCapsHasValueIT.java

```diff
@@ -0,0 +1,306 @@
+
+@ESIntegTestCase.ClusterScope(scope = ESIntegTestCase.Scope.TEST)
+public class FieldCapsHasValueIT extends ESIntegTestCase {
+    private final String INDEX1 = "index-1";
+    private final String INDEX2 = "index-2";
+
+    @Override
+    @Before
+    public void setUp() throws Exception {
+        super.setUp();
+        // Settings s = Settings.builder().put(indexSettings()).put(IndexSettings.INDEX_CHECK_ON_STARTUP.getKey(), false).build();
+        Settings indexSettings = Settings.builder().put(indexSettings()).build();
+        assertAcked(
+            prepareCreate(INDEX1).setWaitForActiveShards(ActiveShardCount.ALL)
+                .setSettings(indexSettings)
+                .setMapping("foo", "type=text", "bar", "type=keyword", "bar-alias", "type=alias,path=bar"),
+            new TimeValue(1, TimeUnit.MINUTES)
+        );
+        assertAcked(
+            prepareCreate(INDEX2).setWaitForActiveShards(ActiveShardCount.ALL).setSettings(indexSettings).setMapping("bar", "type=date"),
+            new TimeValue(1, TimeUnit.MINUTES)
+        );
+    }
+
+    @Override
+    protected boolean addMockHttpTransport() {
+        return false;
+    }
+
+    @Override
+    protected int numberOfShards() {
+        return 1; // TODO-MP find solution to remove this
+    }
+
+    @Override
+    protected int numberOfReplicas() {
+        return 1; // TODO-MP find solution to remove this
+    }
+
+    public void testNoFieldsInEmptyIndex() {
+        FieldCapabilitiesResponse response = client().prepareFieldCaps().setFields("*").setIncludeFieldsWithNoValue(false).get();
+
+        assertIndices(response, INDEX1, INDEX2);
+        // Ensure the response has no mapped fields.
+        assertFalse(response.get().containsKey("foo"));
+        assertFalse(response.get().containsKey("bar"));
+        assertFalse(response.get().containsKey("bar-alias"));
+    }
+
+    public void testOnlyFieldsWithValue() {
+        prepareIndex(INDEX1).setSource("foo", "foo-text").get();
+        refresh(INDEX1); // TODO-MP is really needed or am I missing something?
+
+        FieldCapabilitiesResponse response = client().prepareFieldCaps(INDEX1).setFields("*").setIncludeFieldsWithNoValue(false).get();
+
+        assertIndices(response, INDEX1);
+        assertTrue(response.get().containsKey("foo"));
+        // Check the capabilities for the 'foo' field.
+        Map<String, FieldCapabilities> fooField = response.getField("foo");
+        assertEquals(1, fooField.size());
+        assertTrue(fooField.containsKey("text"));
+        assertEquals(
+            new FieldCapabilities("foo", "text", false, true, false, null, null, null, Collections.emptyMap()),
+            fooField.get("text")
+        );
+    }
+
+    public void testOnlyFieldsWithValueAfterNodesRestart() throws Exception {
+        prepareIndex(INDEX1).setSource("foo", "foo-text").get();
+        internalCluster().fullRestart();
+        ensureGreen(INDEX1);
```

Comment:
Is this the correct way to index a document, restart the cluster and wait for it to be ready (as it should be for search?), asking because it seems that results vary and test is not really stable. Eg. once in 20 runs it fails the assert below (L:101) because there is no field `foo`. 
While some other times it fails to `ensureGreen`, I think I am missing something but I am not sure what.